Added saving to postgres DB capabilities

VladShtompel opened the pull request:
Sorry for the 1 commit, it was worked on in a different fork.

I'll just briefly go over some non-trivial/uncommented changes:

- **contrib/detection_demo/utils.py**
The `non_max_suppression` function for yolo would return `(x1, y1, x2, y2, score, class_conf, class)`
I changed it to instead return  `(x1, y1, x2, y2, score, *all_classes_confs, class)`
This info is necessary for better operations research later, and otherwise has no effect on display or other components. The only effect is that now each prediction is ~12 times bigger in size (86 values instead of 7) which may slow down the read\write a bit. If this turns out to be a real issue I'll make this toggle-able.

- **contrib/yolov3.py** 
Just some minor changes to make it compatible with the changes described above.

- **core/component.py**
This came up when I wanted to use zpc for something and realized that the current approach is to write something along the lines of `self._routines[i].foo()` inside the component, which just looked wrong to me (also kinda error prone).

- **core/message.py**
`OrderdDict` preserves input order, seems fitting for this use case.

- **core/routine.py**
The logging format had some strange dashes, changed them to normal ones.

- **utils/structures/instances.py**
When trying to do something along the lines of this: (assume `predictions` is of type `Instances`)
```python
for pred in predictions:
    # do something..
```
It would crash with an error (`TypeError: len() of a 0-d tensor`).. After some debugging I realized that some values inside Instances look like `tensor(0.2)` instad of `tensor([0.2])`, which caused the error.
I added this patch to fix that. It worked and had no adverse effects in display or any other component.

OmriKaduri commented:
Hi @VladShtompel ,

I have a few questions regarding the design of the `DBHandler`:
1. I guess the motivation behind it is to encapsulate the exception handling of the `commit` function. Although I agree that we should encapsulate the exception handling, I'm not sure if we want to encapsulate the call to `db.session.add(item)`. But, in order to encapsulate only the error handling, we need to find another way to "know when a commit need to be done".
Therefore, I think we need to create a decorator for `after_main_logic` routine, and add this commit-handling part to it. This idea is an inspiration from [this nice blog](https://chase-seibert.github.io/blog/2016/03/31/flask-sqlalchemy-sessionless.html), where he implement exactly that using Flask's `after_request` decorator:
```python
@app.after_request
def session_commit(response):
    if response.status_code >= 400:
        return
    try:
        session.commit()
    except DatabaseError:
        session.rollback()
        raise
```
And I think the call to the `insert` function (or any other function exposed by sqlalchemy: update, delete, etc.) shouldn't be encapsulated.

Generally, I think that `Routine` and `Component` classes should have a couple of decorators to enable flexibility regarding their "Lifecycle" operations.

What do you think?

2. Regarding the try-catch blocks inside the your `insert` function. Aren't all those exceptions are `SQLAlchemyError`? If so, why we need a catch block for every single one, if all will be catched by the single catch block of `SQLAlchemyError`?

VladShtompel commented:
Hi @OmriKaduri, thanks for taking the time to look at this!

> I guess the motivation behind it is to encapsulate the exception handling of the commit function. 

Not strictly. The motivation behind `DBHandler` was to encapsulate **all** database related interactions, this includes definition of tables, connection, insertion and as you mentioned also error handling. 

I read the blog post you linked and I really like the idea of using `declarative_base` for operations such as `save`, `update`, `delete` but I'm unsure about the `after_main_logic` decorator to handle commits part. Maybe I can utilize `Events` for this instead?

> why we need a catch block for every single one, if all will be catched by the single catch block of `SQLAlchemyError`?

As you can see all of the exceptions are currently handled the same way, I simply added the different catch blocks to "prepare the groundwork" in case someone will want custom behavior for the different exceptions in the future.

OmriKaduri commented:
Ok, so personally I think that we should not encapsulate all database related interactions, since they are encapsulated "enough" (IMHO) my the ORM (sqlalchemy). Therefore I still suggest that will wrap only the part of exception handling.
Regarding your note about "preparing the groundwork", I think that in case anyone will want to add special error catch-blocks he'll know what catch block to write, and meanwhile we shouldn't just catch them without a special reason, so you should delete the unnecessary blocks.

VladShtompel commented:
@OmriKaduri 

So basically, if I understood you correctly, practically what you're saying is:
1. Get rid of `DBHandler.py`
2. Move all database interactions to `save_predictions.py`

This means that if someone wants to implement something similar they must rewrite (or at least copy and paste) all the db connection boilerplate and also the `declerative_base` class (if we decide on making one).

Not to imply that this is a big deal, although I do think its a shame to have to rewrite that.

BTW what you're suggesting is what I was about to do from the start, @Batush123 suggested that instead I should abstract away all the database interactions, so I'm really not sure what should I do next about this.

I'd love some input from @ItamarWilf and @ItayHoresh as well.

ItayHoresh commented on pipert/contrib/database/handler.py:
```diff
@@ -0,0 +1,176 @@
+import re
+
+
+class DBHandler:
```
Is this class only for Postgresql? If so, I suggested that the name of the class will be PSQLDBHandler.

ItayHoresh commented on pipert/contrib/database/handler.py:
```diff
@@ -0,0 +1,176 @@
+				self.db.session.close()
+				raise
+
+	def _define_tables(self) -> None:
```
I think that this method needs to abstract and every project will customize his tables.

VladShtompel commented:
Considered the suggestions from @OmriKaduri  and @ItayHoresh and made some changes:

- Renamed the handler to `PSQLDBHandler`.

- The handler now only handles connection, it has an abstract method for defining tables to be implemented individually for each use case.

- Moved committing to db to a function that is used as an `event_handler` for `AFTER_LOGIC` Event

- Removed the extra exception handling.

- Fixed some typos
